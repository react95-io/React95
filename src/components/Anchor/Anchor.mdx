---
name: Anchor
menu: Components
---

import { Playground, Props } from 'docz';

import Anchor from './Anchor';

# Anchor

## Usage

<Playground>
  <Anchor href='https://expensive.toys' target='_blank'>
    Expensive Toys
  </Anchor>
</Playground>

## API

### Props

<Props of={Anchor} />

### Import

```
import { Anchor } from 'react95'
```
