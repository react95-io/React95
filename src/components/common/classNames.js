/**
 * Given a string, number, object, or array, extract
 * the class names configuration from it, and return
 * as a single class string.
 * @param {*} [value]
 * @return {string}
 */
export const getClassesForValue = value => {
  if (value) {
    if (typeof value === 'string' || typeof value === 'number') {
      return value.toString();
    }

    if (Array.isArray(value)) {
      return value.map(getClassesForValue).join(' ');
    }

    if (typeof value === 'object' && !Array.isArray(value)) {
      return Object.keys(value)
        .filter(key => Boolean(value[key]))
        .map(getClassesForValue)
        .join(' ');
    }
  }

  return '';
};

/**
 * Given any number of arguments of type string, number,
 * object, or a multidimensional array, extract a class names
 * configuration from it, and return as a single classes string.
 * For object values, uses keys as classes, but only if their
 * associated values are truthy.
 * @example
 * const result = classNames('alpaca', ['bear', 'cat']);
 * --> 'alpaca bear cat'
 * const result = classNames('alpaca', { bear: null, cat: true });
 * --> 'alpaca cat' // 'bear' is null, so is not included
 * const result = classNames(['alpaca', { bear: true }, ['cat']]);
 * --> 'alpaca bear cat' // multi-dim array is flattened
 * @param {...*} args
 * @return {string}
 */
const classNames = (...args) => {
  return args.reduce((classesString, arg) => {
    const classesForArg = getClassesForValue(arg);

    // leading whitespace is okay as the final string is trimmed
    return `${classesString} ${classesForArg}`.trim();
  }, '');
};

export default classNames;
