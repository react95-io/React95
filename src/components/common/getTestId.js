const getTestId = (testIdPrefix, appendText = '') =>
  testIdPrefix ? `${testIdPrefix}${appendText}`.trim() : undefined;

export default getTestId;
