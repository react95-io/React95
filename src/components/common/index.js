import { css } from "styled-components";

export const shadow = `4px 4px 10px 0 rgba(0,0,0,0.35)`;
export const insetShadow = `inset 3px 3px 10px rgba(0, 0, 0, 0.2)`;

export const createDisabledTextStyles = () => css`
  color: ${({ theme }) => theme.textDisabled};
  text-shadow: 1px 1px ${({ theme }) => theme.textDisabledShadow};
  /* filter: grayscale(100%); */
`;
export const createBoxStyles = () =>
  css`
    box-sizing: border-box;
    display: inline-block;
    background-color: ${({ theme }) => theme.material};
    color: ${({ theme }) => theme.text};
  `;
export const createBorderStyles = (invert = false) =>
  invert
    ? css`
        border-style: solid;
        border-width: 2px;
        border-left-color: ${({ theme }) => theme.borderDarkest};
        border-top-color: ${({ theme }) => theme.borderDarkest};
        border-right-color: ${({ theme }) => theme.borderLightest};
        border-bottom-color: ${({ theme }) => theme.borderLightest};
        box-shadow: ${props => props.shadow && shadow + ", "} inset 1px 1px 0px
            1px ${({ theme }) => theme.borderDark},
          inset -1px -1px 0 1px ${({ theme }) => theme.borderLight};
      `
    : css`
        border-style: solid;
        border-width: 2px;
        border-left-color: ${({ theme }) => theme.borderLightest};
        border-top-color: ${({ theme }) => theme.borderLightest};
        border-right-color: ${({ theme }) => theme.borderDarkest};
        border-bottom-color: ${({ theme }) => theme.borderDarkest};
        box-shadow: ${props => props.shadow && shadow + ", "} inset 1px 1px 0px
            1px ${({ theme }) => theme.borderLight},
          inset -1px -1px 0 1px ${({ theme }) => theme.borderDark};
      `;
