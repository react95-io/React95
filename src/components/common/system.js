// TODO - implement styled-system

export const blockSizes = {
  sm: '27px',
  md: '35px',
  lg: '43px'
};

export const fontFamily = 'sans-serif';
