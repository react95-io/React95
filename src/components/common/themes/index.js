import azureOrange from './azureOrange';
import bee from './bee';
import blackAndWhite from './blackAndWhite';
import brick from './brick';
import candy from './candy';
import coldGray from './coldGray';
import eggplant from './eggplant';
import highContrast from './highContrast';
import lilac from './lilac';
import lilacRoseDark from './lilacRoseDark';
import maple from './maple';
import marine from './marine';
import matrix from './matrix';
import millenium from './millenium';
import modernDark from './modernDark';
import molecule from './molecule';
import ninjaTurtles from './ninjaTurtles';
import olive from './olive';
import original from './original';
import pamelaAnderson from './pamelaAnderson';
import plum from './plum';
import rainyDay from './rainyDay';
import rose from './rose';
import slate from './slate';
import spruce from './spruce';
import theSixtiesUSA from './theSixtiesUSA';
import tokyoDark from './tokyoDark';
import tooSexy from './tooSexy';
import travel from './travel';
import vaporTeal from './vaporTeal';
import vermillion from './vermillion';
import violetDark from './violetDark';
import water from './water';

export default {
  azureOrange,
  bee,
  blackAndWhite,
  brick,
  candy,
  coldGray,
  eggplant,
  highContrast,
  lilac,
  lilacRoseDark,
  maple,
  marine,
  matrix,
  millenium,
  modernDark,
  molecule,
  ninjaTurtles,
  olive,
  original,
  pamelaAnderson,
  plum,
  rainyDay,
  rose,
  slate,
  spruce,
  theSixtiesUSA,
  tokyoDark,
  tooSexy,
  travel,
  vaporTeal,
  vermillion,
  violetDark,
  water
};
