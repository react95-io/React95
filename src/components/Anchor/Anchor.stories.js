import React from 'react';
import styled from 'styled-components';

import Anchor from './Anchor';

const Wrapper = styled.div`
  padding: 5rem;
  background: ${({ theme }) => theme.material};
`;

export default {
  title: 'Anchor',
  component: Anchor,
  decorators: [story => <Wrapper>{story()}</Wrapper>]
};

export const Default = () => (
  <h1>
    Everybody likes{' '}
    <Anchor href='https://expensive.toys' target='_blank'>
      Expensive Toys
    </Anchor>
  </h1>
);

Default.story = {
  name: 'default'
};
