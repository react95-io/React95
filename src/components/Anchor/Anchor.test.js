import React from "react";
import { ThemeProvider } from "styled-components";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import PropTypes from "prop-types";
import { assertPropTypes } from "check-prop-types";

import Anchor from "./Anchor";
import themes from "../common/themes";
import { fontSizes } from "../common/system";

describe("Component: Anchor", () => {

  const testProps = {
    className: "my-class",
    href: "#",
    style: { color: "red" },
    children: "foo",
  }

  it("warns if props are missing or the wrong type", () => {
    expect(() => { assertPropTypes(Anchor.propTypes, { ...testProps, children: undefined}, 'prop', Anchor.children); }).toThrow();
    expect(() => { assertPropTypes(Anchor.propTypes, { ...testProps, href: undefined }, 'prop', Anchor.href); }).toThrow();
    expect(() => { assertPropTypes(Anchor.propTypes, { ...testProps, className: 123 }, 'prop', Anchor.className); }).toThrow();
    expect(() => { assertPropTypes(Anchor.propTypes, { ...testProps, style: "red" }, 'prop', Anchor.style); }).toThrow();
  });

  it("renders", () => {
    const wrapper = shallow(<Anchor />);
    expect(wrapper.find("Anchor__StyledAnchor").length).toBe(1);
  });

  it("renders children", () => {
    const wrapper = shallow(<Anchor>text</Anchor>);
    expect(wrapper.text()).toBe("text");
  });

  it("renders href", () => {
    const wrapper = shallow(<Anchor href="#">text</Anchor>);
    expect(wrapper.props().href).toBe("#");
  });

  it("renders className", () => {
    const wrapper = shallow(<Anchor href="#" className="my-class">text</Anchor>);
    expect(wrapper.props().className).toBe("my-class");
  });

  it("renders style object", () => {
    const wrapper = shallow(
      <Anchor
        href="#"
        className="my-class"
        style={{ color: "blue" }}
      >
        text
      </Anchor>
    );
    expect(wrapper.props().style.color).toBe("blue");
  });

  it("renders other props passed", () => {
    const wrapper = shallow(
      <Anchor
        href="#"
        className="my-class"
        style={{ color: "blue" }}
        a={1}
        b={2}
        c={3}
      >
        text
      </Anchor>
    );
    expect(wrapper.props().a).toBe(1);
    expect(wrapper.props().b).toBe(2);
    expect(wrapper.props().c).toBe(3);
  });

  it("renders default component styles", () => {
    const tree = renderer.create(<Anchor theme={themes.default} href="#">text</Anchor>).toJSON();
    expect(tree).toHaveStyleRule("color", themes.default.anchor);
    expect(tree).toHaveStyleRule("font-size", "inherit");
    expect(tree).toHaveStyleRule("text-decoration", "underline");
    expect(tree).toHaveStyleRule("color", themes.default.anchorVisited, { modifier: ":visited" });
  });

  it("renders correct font size", () => {
    const tree = renderer.create(<Anchor size="sm" />).toJSON();
    expect(tree).toHaveStyleRule("font-size", "15px");
  });

  it("renders matching snapshot", () => {
    const component = renderer.create(
      <Anchor theme={themes.default} href="#">
        text
      </Anchor>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
})
