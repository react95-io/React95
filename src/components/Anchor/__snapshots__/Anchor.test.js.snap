// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Anchor renders matching snapshot 1`] = `
.c0 {
  color: #1034a6;
  font-size: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:visited {
  color: #440381;
}

<a
  className="c0"
  href="#"
>
  text
</a>
`;
